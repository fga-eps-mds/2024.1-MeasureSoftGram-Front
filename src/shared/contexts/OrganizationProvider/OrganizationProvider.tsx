import React, { createContext, useState, useContext, ReactNode, useMemo, useEffect } from 'react';
import { Organization } from '@customTypes/organization';
import { organizationQuery } from '@services/organization';
import { toast } from 'react-toastify';
import { useAuth } from '@contexts/Auth';

interface Props {
  children: ReactNode;
}

interface IOrganizationContext {
  currentOrganization: Organization | null;
  currentOrganizations: Organization[];
  setCurrentOrganizations: (organizations: Organization[]) => void;
  organizationList: Organization[];
  isLoading: boolean;
  fetchOrganizations: (forceFetch?: boolean) => void;
}

const OrganizationContext = createContext<IOrganizationContext | undefined>(undefined);

export function OrganizationProvider({ children }: Props) {
  const { session } = useAuth();
  const [currentOrganizations, setCurrentOrganizations] = useState<Organization[]>([]);
  const [currentOrganization, setCurrentOrganization] = useState<Organization | null>(null);
  const [organizationList, setOrganizationList] = useState<Organization[]>([]);
  const [isLoading, setIsLoading] = useState(true);

  const fetchOrganizations = async (forceFetch?: boolean) => {
    if (!session && !forceFetch) return;
    setIsLoading(true);
    try {
      const result = await organizationQuery.getAllOrganization();
      if (result.type === 'success') {
        const organizations = result.value.map(org => ({
          id: org.id ?? '',
          name: org.name,
          description: org.description ?? '',
          url: org.url ?? '',
          products: org.products ?? [],
          key: org.key ?? ''
        }));
        setOrganizationList(organizations);
      } else {
        toast.error("Erro ao carregar organizações.");
      }
    } catch (error) {
      console.error("Failed to fetch organizations:", error);
      toast.error("Erro ao carregar organizações. Por favor, tente novamente.");
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    fetchOrganizations();
  }, []);

  useEffect(() => {
    if (organizationList.length > 0 && currentOrganizations.length === 0) {
      setCurrentOrganizations([organizationList[0]]);
    }
  }, [organizationList, currentOrganizations]);

  useEffect(() => {
    if (currentOrganizations.length > 0) {
      setCurrentOrganization(currentOrganizations[0]);
    } else {
      setCurrentOrganization(null);
    }
  }, [currentOrganizations]);

  const value = useMemo(() => {
    return {
      currentOrganization,
      currentOrganizations,
      setCurrentOrganizations,
      organizationList,
      isLoading,
      fetchOrganizations
    };
  }, [currentOrganization, currentOrganizations, organizationList, isLoading]);

  return <OrganizationContext.Provider value={value}>{children}</OrganizationContext.Provider>;
}

export function useOrganizationContext() {
  const context = useContext(OrganizationContext);

  if (context === undefined) {
    throw new Error('OrganizationContext must be used within a OrganizationProvider');
  }

  return context;
}
