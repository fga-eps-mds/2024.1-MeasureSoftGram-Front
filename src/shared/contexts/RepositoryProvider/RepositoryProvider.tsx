import React, { createContext, useState, useContext, ReactNode, useMemo } from 'react';

import { Repository, Historical, Result } from '@customTypes/repository';
import { RepositoriesLatestTsqmi, TsqmiValue } from '@customTypes/product';

interface Props {
  children: ReactNode;
}

interface IRepositoryContext {
  currentRepository?: Repository;
  setCurrentRepository: (repository: Repository) => void;
  repositoryList?: Repository[];
  setRepositoryList: (repository: Repository) => void;
  characteristics: string[];
  setCharacteristics: (characteristics: string[]) => void;
  subCharacteristics: string[];
  setSubCharacteristics: (subCharacteristics: string[]) => void;
  measures: string[];
  setMeasures: (measures: string[]) => void;
  metrics: string[];
  setMetrics: (metrics: string[]) => void;
  historicalTSQMI: Historical;
  setHistoricalTSQMI: (historical: Historical) => void;
  latestTSQMI: TsqmiValue;
  setLatestTSQMI: (result: TsqmiValue) => void;
  latestTSQMIBadgeUrl: string;
  setLatestTSQMIBadgeUrl: (result: string) => void;
  repositoriesLatestTsqmi: RepositoriesLatestTsqmi;
  setRepositoriesLatestTsqmi: (result: RepositoriesLatestTsqmi) => void;
}

const RepositoryContext = createContext<IRepositoryContext | undefined>(undefined);

export function RepositoryProvider({ children }: Props) {
  const [currentRepository, setCurrentRepository] = useState<Repository | undefined>();
  const [repositoryList, setRepositoryList] = useState<Repository[]>();

  const [characteristics, setCharacteristics] = useState<string[]>([]);
  const [subCharacteristics, setSubCharacteristics] = useState<string[]>([]);
  const [measures, setMeasures] = useState<string[]>([]);
  const [metrics, setMetrics] = useState<string[]>([]);
  const [historicalTSQMI, setHistoricalTSQMI] = useState<Historical>();
  const [latestTSQMI, setLatestTSQMI] = useState<TsqmiValue>();
  const [latestTSQMIBadgeUrl, setLatestTSQMIBadgeUrl] = useState<string>();
  const [repositoriesLatestTsqmi, setRepositoriesLatestTsqmi] = useState<RepositoriesLatestTsqmi>();


  const value = useMemo(
    () => ({
      currentRepository,
      setCurrentRepository,
      repositoryList,
      setRepositoryList,
      characteristics,
      setCharacteristics,
      subCharacteristics,
      setSubCharacteristics,
      measures,
      setMeasures,
      metrics,
      setMetrics,
      historicalTSQMI,
      setHistoricalTSQMI,
      latestTSQMI,
      setLatestTSQMI,
      latestTSQMIBadgeUrl,
      setLatestTSQMIBadgeUrl,
      repositoriesLatestTsqmi,
      setRepositoriesLatestTsqmi
    }),
    [currentRepository, repositoryList, characteristics, subCharacteristics, measures, metrics, historicalTSQMI, latestTSQMI]
  );

  return <RepositoryContext.Provider value={value}>{children}</RepositoryContext.Provider>;
}

export function useRepositoryContext() {
  const context = useContext(RepositoryContext);

  if (context === undefined) {
    throw new Error('useRepositoryContext must be used within a RepositoryContext');
  }

  return context;
}
