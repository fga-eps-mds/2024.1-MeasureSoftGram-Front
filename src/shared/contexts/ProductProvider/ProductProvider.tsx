import React, { createContext, useState, useContext, ReactNode, useMemo, useCallback, useEffect } from 'react';

import { Product } from '@customTypes/product';
import { productQuery } from '@services/product';
import { useOrganizationContext } from '@contexts/OrganizationProvider';

interface Props {
  children: ReactNode;
}

export interface IProductContext {
  currentProduct?: Product | null;
  setCurrentProduct: (product: Product | null) => void;
  productsList?: Product[];
  updateProductList: (products: Product[]) => void;
}

export const ProductContext = createContext<IProductContext | undefined>(undefined);

export function ProductProvider({ children }: Props) {
  const [currentProduct, setCurrentProduct] = useState<Product | null | undefined>(undefined); // Initialize with undefined
  const [productsList, setProductsList] = useState<Product[]>([]);

  const { currentOrganization } = useOrganizationContext();

  const updateProductList = useCallback((products: Product[]) => {
    setProductsList(products);
  }, []);

  const loadAllProducts = async () => {
    try {
      if (!currentOrganization) {
        updateProductList([]);
        return;
      }

      const result = await productQuery.getAllProducts(currentOrganization.id);
      if (result.status === 200) {
        updateProductList(result.data.results);
      }
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    if (currentOrganization) {
      setCurrentProduct(null);
      loadAllProducts();
    }
  }, [currentOrganization]);

  const value = useMemo(
    () => ({
      currentProduct,
      setCurrentProduct,
      productsList,
      updateProductList,
    }),
    [currentProduct, productsList, updateProductList]
  );

  return <ProductContext.Provider value={value}>{children}</ProductContext.Provider>;
}

export function useProductContext() {
  const context = useContext(ProductContext);

  if (!context) {
    throw new Error('useProductContext must be used within a ProductContext');
  }

  return context;
}
