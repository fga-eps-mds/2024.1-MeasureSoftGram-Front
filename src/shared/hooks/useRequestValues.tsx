import { useOrganizationContext } from '@contexts/OrganizationProvider';
import { useProductContext } from '@contexts/ProductProvider';
import { useRepositoryContext } from '@contexts/RepositoryProvider';
import useSWR from 'swr';
import { Historical } from '@customTypes/repository';
import api from '@services/api';
import _ from 'lodash';
import { productQuery } from '@services/product';
import useBoolean from './useBoolean';

interface Props {
  type: 'historical-values' | 'latest-values';
  value: 'characteristics' | 'subcharacteristics' | 'measures' | 'metrics';
  addHistoricalTSQMI?: boolean;
  addCurrentGoal?: boolean;
  metricsSource?: 'github' | 'sonarqube';
}

export function useRequestValues({
  type,
  value,
  addHistoricalTSQMI = false,
  addCurrentGoal = false,
  metricsSource
}: Props) {
  const { currentOrganization } = useOrganizationContext();
  const { currentProduct } = useProductContext();
  const { currentRepository, historicalTSQMI } = useRepositoryContext();

  const { value: isLoading, setTrue: setLoading, setFalse: setIsLoadingEnd } = useBoolean(false);

  const { data, error, isValidating } = useSWR<{ results: Historical[] }>(
    `organizations/${currentOrganization?.id}` +
    `/products/${currentProduct?.id}` +
    `/repositories/${currentRepository?.id}` +
    `/${type}/${value}/`,
    (url) => {
      setLoading();

      return api
        .get(url)
        .then(async (response) => {
          if (addCurrentGoal && currentOrganization?.id && currentProduct?.id) {
            try {
              const { data: currentGoal } = await productQuery.getCurrentReleaseGoal(
                currentOrganization.id,
                currentProduct.id
              );
              response.data?.results?.forEach((res: Historical) => {
                res.goal = currentGoal.data[res.key];
              });
            } catch {
              response.data?.results?.forEach((res: Historical) => {
                res.goal = undefined;
              });
            }
          }

          const data = response.data;

          return data;
        })
        .finally(() => {
          setIsLoadingEnd();
        });
    },
    {
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
      refreshWhenOffline: false,
      refreshWhenHidden: false,
      shouldRetryOnError: false
    }
  );


  let returnData = _.cloneDeep(data?.results ?? []);

  if (addHistoricalTSQMI && !_.isEmpty(returnData) && !_.find(returnData, { key: 'TSQMI' })) {
    returnData.push(historicalTSQMI);
  }

  if (returnData?.length && metricsSource) {
    if (metricsSource === 'github') {
      returnData = returnData.filter(
        (res: any) => res.key === 'ci_feedback_time' || res.key === 'closed_issues' || res.key === 'total_issues'
      );
    }

    if (metricsSource === 'sonarqube') {
      returnData = returnData.filter(
        (res: any) => res.key !== 'ci_feedback_time' && res.key !== 'closed_issues' && res.key !== 'total_issues'
      );

      console.log(returnData);
    }
  }

  console.log(metricsSource, returnData);
  return {
    data: returnData,
    error,
    isLoading,
    isValidating,
    isEmpty: data?.results.length === 0 && !isLoading && !error && !isValidating
  };
}
