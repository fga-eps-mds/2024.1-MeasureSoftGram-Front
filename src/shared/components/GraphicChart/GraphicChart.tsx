import React, { useMemo, useState } from 'react';
import ReactEcharts from 'echarts-for-react';

import formatCharacteristicsHistory from '@utils/formatCharacteristicsHistory';
import formatMsgramChart from '@utils/formatMsgramChart';
import formatRadarChart from '@utils/formatRadarChart';
import formatGaugeChart from '@utils/formatGaugeChart';
import { Alert, Box, Fade, IconButton, Skeleton } from '@mui/material';
import { useRequestValues } from '@hooks/useRequestValues';
import { Historical } from '@customTypes/repository';
import _ from 'lodash';
import { useProductConfigFilterContext } from '@contexts/ProductConfigFilterProvider/ProductConfigFilterProvider';
import KeyboardDoubleArrowDownIcon from '@mui/icons-material/KeyboardDoubleArrowDown';
import KeyboardDoubleArrowUpIcon from '@mui/icons-material/KeyboardDoubleArrowUp';
import { useProductContext } from '@contexts/ProductProvider';

interface Prop {
  title: string;
  type: 'line' | 'msg' | 'radar' | 'gauge';
  value: 'characteristics' | 'subcharacteristics' | 'measures' | 'metrics';
  valueType?: 'historical-values' | 'latest-values';
  autoGrid?: boolean;
  addHistoricalTSQMI?: boolean;
  addCurrentGoal?: boolean;
  metricsSource?: 'github' | 'sonarqube'
}

type formatFunctionType = {
  [key: string]: Function;
};

const chartOption: formatFunctionType = {
  line: formatCharacteristicsHistory,
  msg: formatMsgramChart,
  radar: formatRadarChart,
  gauge: formatGaugeChart
};

const GraphicChart = ({
  title,
  type,
  value,
  valueType = 'historical-values',
  autoGrid = false,
  addHistoricalTSQMI = false,
  addCurrentGoal = false,
  metricsSource,
}: Prop) => {
  const {
    data: historical,
    error,
    isLoading,
    isEmpty
  } = useRequestValues({ type: valueType, value, addHistoricalTSQMI, addCurrentGoal, metricsSource });
  const { hasKey } = useProductConfigFilterContext();
  const [showCharts, setShowCharts] = useState(false);
  const { currentProduct } = useProductContext();


  const sliceHistorical = (rowIdx: number): Historical[] => {
    if (!autoGrid) return historical;
    return historical.slice(numChartsPerLine * rowIdx, numChartsPerLine * (rowIdx + 1));
  };

  const historicalLength: number = historical?.length ?? 0;
  const numChartsPerLine: number = 2;
  const numLines: number = !autoGrid ? 1 : Math.ceil(historicalLength / numChartsPerLine);

  let chartBoxHeight: string = 'auto';
  let chartStyle: React.CSSProperties = {};

  if (type === 'msg') {
    chartBoxHeight = `${historicalLength * 82 + 85}px`;
    chartStyle = { height: chartBoxHeight };
  }

  const chartsOption = useMemo(
    () =>
      _.range(numLines).map((i) => ({
        ...chartOption[type]({
          historical: _.filter(sliceHistorical(i), (item) => metricsSource === 'github' || hasKey(item.key)),
          title: i === 0 ? title : '',
          isEmpty: isEmpty || error,
          redLimit: currentProduct?.gaugeRedLimit,
          yellowLimit: currentProduct?.gaugeYellowLimit
        }),
        key: `graphic-chart-${i}`
      })
      ),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [historical, title, isEmpty, error]
  );

  const filteredChartsOptions = chartsOption.filter((option, index) => {
    if (type === 'gauge') {
      return (index <= 1) ? true : false;
    }
    return true;
  });

  return isLoading ? (
    <Skeleton variant="rectangular" height="300px" />
  ) : (
    <>
      <Fade in timeout={1500}>
        <Box
          bgcolor="white"
          borderRadius="4px"
          boxShadow="0px 2px 1px -1px rgba(0,0,0,0.2), 0px 1px 1px 0px rgba(0,0,0,0.14), 0px 1px 3px 0px rgba(0,0,0,0.12)"
          paddingX="20px"
          paddingY="10px"
          width="100%"
          height={chartBoxHeight}
        >
          {(type !== 'gauge') || (type === 'gauge' && showCharts) ?
            chartsOption.map((option) => (
              <>
                <ReactEcharts key={option.key} notMerge lazyUpdate style={chartStyle} option={option} />
              </>
            ))
            :
            filteredChartsOptions.map((option) => (
              <>
                <ReactEcharts key={option.key} notMerge lazyUpdate style={chartStyle} option={option} />
              </>
            ))
          }

          {(type === 'gauge' && chartsOption.length > 2) &&
            <Box
              width="100%"
              display="flex"
              flexDirection="row"
              justifyContent="center"
            >
              <IconButton onClick={() => { setShowCharts(prev => !prev) }}>
                {showCharts ? <KeyboardDoubleArrowUpIcon fontSize="large" /> : <KeyboardDoubleArrowDownIcon fontSize="large" />}
              </IconButton>
            </Box>
          }
        </Box>
      </Fade>
      {error && (
        <Fade in timeout={1000}>
          <Alert severity="error">Ocorreu um erro ao tentar carregar as informações</Alert>
        </Fade>
      )}
      {isEmpty && (
        <Fade in timeout={1000}>
          <Alert variant="standard" severity="warning">
            Não há dados para serem exibidos
          </Alert>
        </Fade>
      )}
    </>
  );
};

export default GraphicChart;
