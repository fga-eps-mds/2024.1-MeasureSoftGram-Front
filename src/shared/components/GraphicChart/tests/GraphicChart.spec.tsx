import React from 'react';
import { OrganizationProvider } from '@contexts/OrganizationProvider';
import { ProductProvider } from '@contexts/ProductProvider';
import { RepositoryProvider } from '@contexts/RepositoryProvider';
import { fireEvent, render } from '@testing-library/react';
import { useRequestValues } from '@hooks/useRequestValues';
import GraphicChart from '../GraphicChart';

interface Props {
  children: React.ReactNode;
}

const AllTheProviders = ({ children }: Props) => (
  <OrganizationProvider>
    <ProductProvider>
      <RepositoryProvider>{children}</RepositoryProvider>
    </ProductProvider>
  </OrganizationProvider>
);

jest.mock('@hooks/useRequestValues');

const data = [
  {
    id: 1,
    key: 'reliability',
    name: 'Reliability',
    description: null,
    history: [
      {
        id: 286,
        characteristic_id: 1,
        value: 0.9999996641383778,
        created_at: '2023-06-20T22:02:20.715198-03:00'
      },
      {
        id: 288,
        characteristic_id: 1,
        value: 0.9999999801554467,
        created_at: '2023-06-20T22:07:18.486582-03:00'
      }
    ]
  },
  {
    id: 2,
    key: 'maintainability',
    name: 'Maintainability',
    description: null,
    history: [
      {
        id: 287,
        characteristic_id: 2,
        value: 0.4036922248628375,
        created_at: '2023-06-20T22:02:20.715198-03:00'
      },
      {
        id: 289,
        characteristic_id: 2,
        value: 0.7139040869240877,
        created_at: '2023-06-20T22:07:18.486582-03:00'
      }
    ]
  }
];

const measureData = [
  {
    id: 1,
    key: "passed_tests",
    name: "Passed Tests",
    description: null,
    latest: {
      id: 318,
      measure_id: 1,
      value: 0.6907459278269361,
      created_at: "2023-10-13T13:09:48-03:00"
    }
  },
  {
    id: 2,
    key: "test_builds",
    name: "Test Builds",
    description: null,
    latest: {
      id: 372,
      measure_id: 2,
      value: 0.06486645104021993,
      created_at: "2023-10-11T11:55:23-03:00"
    }
  },
  {
    id: 3,
    key: "test_coverage",
    name: "Test Coverage",
    description: null,
    latest: {
      id: 446,
      measure_id: 3,
      value: 0.8674433534143255,
      created_at: "2023-10-08T17:13:06-03:00"
    }
  },
  {
    id: 4,
    key: "non_complex_file_density",
    name: "Non Complex File Density",
    description: null,
    latest: {
      id: 476,
      measure_id: 4,
      value: 0.856364236489889,
      created_at: "2023-10-11T04:21:35-03:00"
    }
  },
  {
    id: 5,
    key: "commented_file_density",
    name: "Commented File Density",
    description: null,
    latest: {
      id: 546,
      measure_id: 5,
      value: 0.8074273827928293,
      created_at: "2023-10-14T06:50:10-03:00"
    }
  },
  {
    id: 6,
    key: "duplication_absense",
    name: "Duplication Absense",
    description: null,
    latest: {
      id: 596,
      measure_id: 6,
      value: 0.5209101386167142,
      created_at: "2023-10-14T07:33:23-03:00"
    }
  }
];

beforeEach(() => {
  jest.clearAllMocks();
});

describe('<GraphicChart />', () => {
  it('should render correctly', () => {
    useRequestValues.mockReturnValue({
      data,
      error: undefined,
      isLoading: false,
      isValidating: false,
      isEmpty: false
    });

    const { container } = render(<GraphicChart type="line" title="title" value="characteristics" />, {
      wrapper: AllTheProviders
    });

    container.firstChild?.firstChild?.setAttribute('_echarts_instance_', 'ec_123');
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with autoGrid', () => {
    useRequestValues.mockReturnValue({
      data,
      error: undefined,
      isLoading: false,
      isValidating: false,
      isEmpty: false
    });

    const { container } = render(<GraphicChart type="line" title="title" value="characteristics" autoGrid />, {
      wrapper: AllTheProviders
    });

    container.firstChild?.firstChild?.setAttribute('_echarts_instance_', 'ec_123');
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with historical TSQMI', () => {
    useRequestValues.mockReturnValue({
      data,
      error: undefined,
      isLoading: false,
      isValidating: false,
      isEmpty: false
    });

    const { container } = render(
      <GraphicChart type="line" title="title" value="characteristics" addHistoricalTSQMI />,
      {
        wrapper: AllTheProviders
      }
    );

    container.firstChild?.firstChild?.setAttribute('_echarts_instance_', 'ec_123');
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with empty data', () => {
    useRequestValues.mockReturnValue({
      data: [],
      error: undefined,
      isLoading: false,
      isValidating: false,
      isEmpty: true
    });

    const { container } = render(<GraphicChart type="line" title="title" value="characteristics" />, {
      wrapper: AllTheProviders
    });

    container.firstChild?.firstChild?.setAttribute('_echarts_instance_', 'ec_123');
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with error', () => {
    useRequestValues.mockReturnValue({
      data: [],
      error: new Error('error'),
      isLoading: false,
      isValidating: false,
      isEmpty: false
    });

    const { container } = render(<GraphicChart type="line" title="title" value="characteristics" />, {
      wrapper: AllTheProviders
    });

    container.firstChild?.firstChild?.setAttribute('_echarts_instance_', 'ec_123');
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with loading', () => {
    useRequestValues.mockReturnValue({
      data: [],
      error: undefined,
      isLoading: true,
      isValidating: false,
      isEmpty: false
    });

    const { container } = render(<GraphicChart type="msg" title="title" value="characteristics" />, {
      wrapper: AllTheProviders
    });

    container.firstChild?.firstChild?.setAttribute('_echarts_instance_', 'ec_123');
    expect(container).toMatchSnapshot();
  });

  it('should filter charts options correctly for gauge type', () => {
    useRequestValues.mockReturnValue({
      data: measureData,
      error: undefined,
      isLoading: false,
      isValidating: false,
      isEmpty: false
    });

    const { container } = render(
      <GraphicChart type="gauge" title="title" autoGrid value="measures" valueType="latest-values" addCurrentGoal />, {
      wrapper: AllTheProviders
    });


    expect(container.getElementsByClassName('echarts-for-react ').length).toBe(2);


  });

  it('should change state when button clicks', async () => {
    useRequestValues.mockReturnValue({
      data: measureData,
      error: undefined,
      isLoading: false,
      isValidating: false,
      isEmpty: false
    });

    const { container, findByRole } = render(
      <GraphicChart type="gauge" title="title" autoGrid value="measures" valueType="latest-values" addCurrentGoal />, {
      wrapper: AllTheProviders
    });

    const button = await findByRole("button");
    fireEvent.click(button);

    expect(container.getElementsByClassName('echarts-for-react ').length).toBe(3);


  });
});
