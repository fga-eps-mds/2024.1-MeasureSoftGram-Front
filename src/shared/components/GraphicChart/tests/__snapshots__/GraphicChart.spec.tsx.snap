// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GraphicChart /> should render correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-13kcqfm"
    style="opacity: 1; webkit-transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      _echarts_instance_="ec_123"
      class="echarts-for-react "
      size-sensor-id="1"
      style="height: 300px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
    >
      <div
        style="position: relative; width: 0px; height: 300px; padding: 0px; margin: 0px; border-width: 0px;"
      />
      <object
        class="size-sensor-object"
        data="about:blank"
        style="display: block; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1; opacity: 0;"
        tabindex="-1"
        type="text/html"
      />
    </div>
  </div>
</div>
`;

exports[`<GraphicChart /> should render correctly with autoGrid 1`] = `
<div>
  <div
    class="MuiBox-root css-13kcqfm"
    style="opacity: 1; webkit-transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      _echarts_instance_="ec_123"
      class="echarts-for-react "
      size-sensor-id="2"
      style="height: 300px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
    >
      <div
        style="position: relative; width: 0px; height: 300px; padding: 0px; margin: 0px; border-width: 0px;"
      />
      <object
        class="size-sensor-object"
        data="about:blank"
        style="display: block; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1; opacity: 0;"
        tabindex="-1"
        type="text/html"
      />
    </div>
  </div>
</div>
`;

exports[`<GraphicChart /> should render correctly with empty data 1`] = `
<div>
  <div
    class="MuiBox-root css-13kcqfm"
    style="opacity: 1; webkit-transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      _echarts_instance_="ec_123"
      class="echarts-for-react "
      size-sensor-id="4"
      style="height: 300px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
    >
      <div
        style="position: relative; width: 0px; height: 300px; padding: 0px; margin: 0px; border-width: 0px;"
      />
      <object
        class="size-sensor-object"
        data="about:blank"
        style="display: block; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1; opacity: 0;"
        tabindex="-1"
        type="text/html"
      />
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorWarning MuiAlert-standardWarning MuiAlert-standard css-y8tvd1-MuiPaper-root-MuiAlert-root"
    role="alert"
    style="opacity: 1; webkit-transition: opacity 1000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 1000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
        data-testid="ReportProblemOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
    >
      Não há dados para serem exibidos
    </div>
  </div>
</div>
`;

exports[`<GraphicChart /> should render correctly with error 1`] = `
<div>
  <div
    class="MuiBox-root css-13kcqfm"
    style="opacity: 1; webkit-transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      _echarts_instance_="ec_123"
      class="echarts-for-react "
      size-sensor-id="5"
      style="height: 300px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
    >
      <div
        style="position: relative; width: 0px; height: 300px; padding: 0px; margin: 0px; border-width: 0px;"
      />
      <object
        class="size-sensor-object"
        data="about:blank"
        style="display: block; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1; opacity: 0;"
        tabindex="-1"
        type="text/html"
      />
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorError MuiAlert-standardError MuiAlert-standard css-x7o5da-MuiPaper-root-MuiAlert-root"
    role="alert"
    style="opacity: 1; webkit-transition: opacity 1000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 1000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
        data-testid="ErrorOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
    >
      Ocorreu um erro ao tentar carregar as informações
    </div>
  </div>
</div>
`;

exports[`<GraphicChart /> should render correctly with historical TSQMI 1`] = `
<div>
  <div
    class="MuiBox-root css-13kcqfm"
    style="opacity: 1; webkit-transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      _echarts_instance_="ec_123"
      class="echarts-for-react "
      size-sensor-id="3"
      style="height: 300px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
    >
      <div
        style="position: relative; width: 0px; height: 300px; padding: 0px; margin: 0px; border-width: 0px;"
      />
      <object
        class="size-sensor-object"
        data="about:blank"
        style="display: block; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1; opacity: 0;"
        tabindex="-1"
        type="text/html"
      />
    </div>
  </div>
</div>
`;

exports[`<GraphicChart /> should render correctly with loading 1`] = `
<div>
  <span
    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-153n0d3-MuiSkeleton-root"
    style="height: 300px;"
  />
</div>
`;
